addon/visit.js

```diff
@@ -117,29 +117,52 @@ class PageVisitor {
         windowId = (await browser.windows.create({url: '/pages/blank.html'})).id;
     }
 
+    // Start processing by triggering ``maxTabs`` next events. This will load
+    // a URL for each event and begin visiting pages in parallel.
+    async start(event) {
+        const windowId = event.detail;
+        for (let i = 0; i < this.maxTabs; i++) {
+            this.doc.dispatchEvent(new CustomEvent(
+              'fathom:next',
+              {detail: windowId}
+            ));
+        }
+    }
+
     // Load next tab from this.urls, or close the window if we're done.
     async next(event) {
         const windowId = event.detail;
 
-        this.urlIndex++;
-        if (this.urlIndex >= this.urls.length) {
-            // Do final cleanup.
-            await this.processAtEndOfRun();
-            browser.tabs.onUpdated.removeListener(this.tabUpdateListener);
-            this.tabUpdateListener = undefined;
-            browser.windows.remove(windowId);
-            this.doc.getElementById('freeze').disabled = false;
-            return;
+        if (this.urlIndex < this.urls.length - 1) {
+            this.urlIndex++;
+            // Capture the current urlIndex here so we can rely on its value in
+            // the success callback for ``browser.tabs.create()``.
+            const urlIndexForMap = this.urlIndex;
+            // Create a new tab with the current url.
+            // The tabs.onUpdated handler in visitAllPages() will dispatch a fathom:freeze
+            // event when the tab has completed loading.
+            browser.tabs.create({
+                windowId: windowId,
+                url: this.urls[this.urlIndex].url,
+                active: false,
+            }).then(tab => {
+                this.tabIdToUrlsIndex.set(tab.id, urlIndexForMap);
+                this.setCurrentStatus({message: 'loading', index: tab.id});
+            }).catch(error => {
+                // TODO: How to handle a failure in loading the tab since we won't have a status item to display an error for?
+                error.log(error);
```

Comment:
Do you mean `console.error` here?

I'm not sure what to do here either, though it doesn't appear to be a common error. The logging is good enough for now, especially if we are moving towards a headless solution in `fathom-vectorize`.

Reply:
Yup. Thanks! We could add another field to the page for displaying errors like this.